### PR TITLE
Add hardware tests for nav env

### PR DESCRIPTION
**Pull request by KavitShah1998**

### CODE REVIEW

**Comment by KavitShah1998** [comment]
For conversation history, please follow this closed PR - https://github.com/facebookresearch/spot-sim2real/pull/43

**Review by akshararai — APPROVED** [review]
Overall looks good to me. Some comments on adding docstrings. But good otherwise.
Test on another robot (maybe @jimmytyyang can test in NYC) before merging.

File: spot_rl_experiments/spot_rl/envs/lang_env.py [hunk 1]
```diff
@@ -318,7 +318,9 @@ def step(self, base_action, arm_action, *args, **kwargs):
         if self.grasp_attempted and not self.navigating_to_place:
             # Determine where to go based on what object we've just grasped
             waypoint_name = rospy.get_param("/viz_place")
-            waypoint = nav_target_from_waypoints(waypoint_name)
+
+            # TEST THIS STEPPPPPPPP!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
[the remark this answers is not in the document]
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/lang_env.py:322]: Actually I did test this. Will remove. Haha thanks @akshararai

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 2]
```diff
@@ -4,61 +4,109 @@
...
 
 DOCK_ID = int(os.environ.get("SPOT_DOCK_ID", 520))
 
 
-def main(spot):
+def parse_arguments(args=sys.argv[1:]):
     parser = get_default_parser()
     parser.add_argument("-g", "--goal")
-    parser.add_argument("-w", "--waypoint")
+    parser.add_argument("-w", "--waypoints")
```
[the remark this answers is not in the document]
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py:29]: This is what we added in the PR for waypoint recording

```
def parse_arguments(args):
    parser = get_default_parser()
    parser.add_argument("-c", "--clutter", help="input:string -> clutter target name")
    parser.add_argument("-p", "--place-target", help="input:string -> place target name")
    parser.add_argument("-n", "--nav-only", help="input:string -> nav target name")
    parser.add_argument("-x", "--create-file", action="store_true", help="input: not needed -> create a new waypoints.yaml file")
    args = parser.parse_args(args=args)

    return args
```

Does this look better? It also mentions the type of input to be specified.

File: tests/hardware_tests/test_nav_env.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    steps_list_traj_set = np.array(steps_list_traj_set)
+
+    # Calculate mean and standard deviation of steps taken to reach each waypoint for all trajectories (size=3 as 3 trajectories in this test)
+    avg_steps_list_traj = np.mean(steps_list_traj_set, axis=0)
+    std_steps_list_traj = np.std(steps_list_traj_set, axis=0)
+
+    return avg_steps_list_traj, std_steps_list_traj
+
+
+def extract_goal_poses_timestamps_steps_from_traj(traj):
```
[the remark this answers is not in the document]
Comment (KavitShah1998) [reply tests/hardware_tests/test_nav_env.py:94]: added

File: tests/hardware_tests/test_nav_env.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    return config
+
+
+def compute_avg_and_std_time(ref_traj_set):
+    # List of time taken to reach each waypoint for each trajectory in ref_traj_set
+    time_list_traj_set = []  # type: List[List[Dict]]
+    for ref_traj in ref_traj_set:
+        # List of time taken to reach each waypoint for a single trajectory in ref_traj; in this test, it will contain 3 elements
+        time_list_traj = []  # type: List[Dict]
+        prev_time = 0
```
[the remark this answers is not in the document]
Comment (KavitShah1998) [reply tests/hardware_tests/test_nav_env.py:49]: The first timestamp does not have 0.0

Thanks for catching this @akshararai . this is a bug.

File: tests/hardware_tests/data/temp_recordings/output.wav
[the remark this answers is not in the document]
Comment (KavitShah1998) [reply tests/hardware_tests/data/temp_recordings/output.wav:1]: removed

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 5]
```diff
@@ -4,61 +4,109 @@
...
-                    time.sleep(0.1)
-    finally:
-        spot.power_off()
+    parser.add_argument("-rt", "--record_trajectories", action="store_true")
+    parser.add_argument("-st", "--save_trajectories")
+    args = parser.parse_args(args=args)
+    return args
+
+
+class WaypointController:
```
[the remark this answers is not in the document]
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py:37]: added

File: spot_rl_experiments/spot_rl/envs/nav_env.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py:29]: updated

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 6]
```diff
@@ -88,6 +136,60 @@ def get_observations(self):
...
+        assert args.goal is not None
+        goal_x, goal_y, goal_heading = [float(i) for i in args.goal.split(",")]
+        nav_targets = [(goal_x, goal_y, goal_heading)]
+
+    # Default value for `args.save_trajectories` is None. Raise error to ask for correct location
+    if (args.save_trajectories is not None) and (
+        not os.path.isdir(args.save_trajectories)
+    ):
+        raise Exception(
+            f"The path for saving trajectories at {args.save_trajectories} either not specified or incorrect. Please provide a correct path"
```
[the remark this answers is not in the document]
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py:169]: I don't think it is a good idea to write a "hard-coded path" like a magic number.
And besides this path might change for every user depending on where the repo is.

Let's hold on this until we make a PR where we are creating a file that generates file paths for each repository.

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 7]
```diff
@@ -25,21 +25,63 @@
...
+        "-d",
+        "--dock",
+        action="store_true",
+        help="input: not needed -> should dock the robot after navigation",
+    )
+    parser.add_argument(
+        "-rt",
+        "--record_trajectories",
+        action="store_true",
+        help="input: not needed -> should record robot's trajectory",
```
Comment (akshararai) [inline spot_rl_experiments/spot_rl/envs/nav_env.py:46]:
```suggestion
        help="input:bool -> True/False ",
```

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 8]
```diff
@@ -25,21 +25,63 @@
...
+    parser.add_argument(
+        "-w",
+        "--waypoints",
+        help="input:string -> nav target waypoints (comma seperated) where robot needs to navigate to",
+    )
+    parser.add_argument(
+        "-d",
+        "--dock",
+        action="store_true",
+        help="input: not needed -> should dock the robot after navigation",
```
Comment (akshararai) [inline spot_rl_experiments/spot_rl/envs/nav_env.py:40]:
```suggestion
        help="input:boolean -> True/False",
```

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 9]
```diff
@@ -25,21 +25,63 @@
...
+        help="input:string -> path to save robot's trajectory",
+    )
     args = parser.parse_args(args=args)
+
     return args
 
 
 class WaypointController:
+    """
+    WaypointController is used to navigate the robot to a given waypoint.
```
Comment (akshararai) [inline spot_rl_experiments/spot_rl/envs/nav_env.py:60]: nice!

File: spot_rl_experiments/spot_rl/envs/nav_env.py [hunk 10]
```diff
@@ -25,21 +25,63 @@
...
+    )
+    parser.add_argument(
+        "-rt",
+        "--record_trajectories",
+        action="store_true",
+        help="input: not needed -> should record robot's trajectory",
+    )
+    parser.add_argument(
+        "-st",
+        "--save_trajectories",
```
Comment (akshararai) [inline spot_rl_experiments/spot_rl/envs/nav_env.py:50]:
```suggestion
        "--save_trajectories_path",
```

File: tests/hardware_tests/test_nav_env.py [hunk 11]
```diff
@@ -41,20 +47,29 @@ def init_config():
 
 
 def compute_avg_and_std_time(ref_traj_set):
+    """
+    Compute average and standard deviation of time taken to reach each waypoint for all trajectories in ref_traj_set
+
+    Args:
+        ref_traj_set: List(List(List(Dict))) where each Dict is a waypoint
```
Comment (akshararai) [inline tests/hardware_tests/test_nav_env.py:54]: what are the different dimensions of this list?

Comment (akshararai) [reply tests/hardware_tests/test_nav_env.py:54]: I prefer just saying list of recorded trajectories :)

**Review by akshararai — APPROVED** [review]

File: spot_rl_experiments/spot_rl/envs/nav_env.py (hunk 7 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py:46]: We don't need any input for this from the user.
We are storing them with "store_true" action.

File: spot_rl_experiments/spot_rl/envs/nav_env.py (hunk 8 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py:40]: We don't need any input for this from the user.
We are storing them with "store_true" action.

File: spot_rl_experiments/spot_rl/envs/nav_env.py (hunk 10 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py]: addressed

File: tests/hardware_tests/test_nav_env.py (hunk 11 above)
Comment (KavitShah1998) [reply tests/hardware_tests/test_nav_env.py:55]: Dict = waypoint (timestamp, pose)
List(Dict) = trajectory for each goal waypoint
List(List(Dict)) = trajectory list for all "n" waypoints robot was asked to go to
List(List(List(Dict))) = aggregation of 15 reference trajectories of type List(List(Dict)) as a List

Comment (KavitShah1998) [reply tests/hardware_tests/test_nav_env.py]: added

File: spot_rl_experiments/spot_rl/envs/nav_env.py (hunk 7 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py]: addressed

File: spot_rl_experiments/spot_rl/envs/nav_env.py (hunk 8 above)
Comment (KavitShah1998) [reply spot_rl_experiments/spot_rl/envs/nav_env.py]: addressed

**Comment by KavitShah1998** [comment]
The PR looks good. Let's merge.

**Review by jimmytyyang — APPROVED** [review]
LGTM!

**Pull request merged**